ext/drreg/drreg.c

```diff
@@ -545,6 +884,57 @@ drreg_event_bb_insert_late(void *drcontext, void *tag, instrlist_t *bb, instr_t
     }
 
     /* After each app write, update spilled app values: */
+#ifdef X86
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++) {
+        if (pt->simd_reg[SIMD_IDX(reg)].in_use) {
+            void *state =
+                drvector_get_entry(&pt->simd_reg[SIMD_IDX(reg)].live, pt->live_idx - 1);
+            reg_id_t spilled_reg = pt->simd_slot_use[pt->simd_reg[SIMD_IDX(reg)].slot];
+            ASSERT(spilled_reg != DR_REG_NULL, "invalid spilled reg");
+
+            if (instr_writes_to_reg(inst, reg, DR_QUERY_INCLUDE_ALL) &&
+                /* Don't bother if reg is dead beyond this write */
+                (ops.conservative || pt->live_idx == 0 ||
+                 !((reg_is_strictly_xmm(spilled_reg) && state >= SIMD_XMM_DEAD &&
+                    state <= SIMD_ZMM_DEAD) ||
+                   (reg_is_strictly_ymm(spilled_reg) && state >= SIMD_YMM_DEAD &&
+                    state <= SIMD_ZMM_DEAD) ||
+                   (reg_is_strictly_zmm(spilled_reg) && state == SIMD_ZMM_DEAD)))) {
+
+                uint tmp_slot = MAX_SIMD_SPILLS;
+                if (!restored_for_simd_read[SIMD_IDX(reg)]) {
+                    tmp_slot = find_simd_free_slot(pt);
+                    if (tmp_slot == MAX_SIMD_SPILLS) {
+                        drreg_report_error(DRREG_ERROR_OUT_OF_SLOTS,
+                                           "failed to preserve tool val wrt app write");
+                    }
+                    spill_reg_indirectly(drcontext, pt, spilled_reg, tmp_slot, bb, inst);
+                }
+
+                instr_t *where =
+                    restored_for_simd_read[SIMD_IDX(reg)] ? instr_get_prev(next) : next;
+                spill_reg_indirectly(drcontext, pt, spilled_reg,
+                                     pt->simd_reg[SIMD_IDX(reg)].slot, bb, where);
+                pt->simd_reg[SIMD_IDX(reg)].ever_spilled = true;
+                if (!restored_for_simd_read[SIMD_IDX(reg)]) {
+                    restore_reg_indirectly(drcontext, pt, spilled_reg, tmp_slot, bb,
+                                           next /*after*/, true);
+                }
+            }
+        } else if (!pt->simd_reg[SIMD_IDX(reg)].native &&
+                   instr_writes_to_reg(inst, reg, DR_QUERY_INCLUDE_ALL)) {
+            /* For an unreserved reg that's written, just drop the slot, even
+             * if it was spilled at an earlier reservation point.
+             */
+            if (pt->simd_reg[SIMD_IDX(reg)].ever_spilled)
+                pt->simd_reg[SIMD_IDX(reg)].ever_spilled = false; /* no need to restore */
+            res = drreg_restore_reg_now(drcontext, bb, inst, pt, reg);
+            if (res != DRREG_SUCCESS)
+                drreg_report_error(res, "slot release on app write failed");
+            pt->simd_pending_unreserved--;
+        }
+    }
```

Comment:
This function is now very large.  Maybe move sthg (this SIMD code?) into helper functions?

Reply:
Yes, helper functions should be fine. @johnfxgalea , this should be ok, right?

Reply:
Same thing here.